File: HashTable/349_IntersectionOfTwoArrays.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-07-02 10:30:33
 */

class Solution {
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        set<int> sets1(nums1.begin(), nums1.end());
        vector<int> ans;
        for(auto &n : nums2){
            if(sets1.count(n)){
                ans.push_back(n);
                sets1.erase(n);
            }
        }
        return ans;
    }
};

class Solution_2{
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());
        int i = 0, j=0;
        vector<int> ans;
        while(i < nums1.size() && j < nums2.size()){
            if(nums1[i] > nums2[j]){
                j++;
            }
            else if(nums1[i] < nums2[j]){
                i++;
            }
            else{
                if(ans.size()==0 || nums1[i] != ans[ans.size()-1]){
                    ans.push_back(nums1[i]);
                }
                i++;
                j++;
            }
        }
        return ans;
    }
};

/*
[]
[]
[-1]
[]
[1,2,2,2,10,5]
[1,3,4,5,9]
*/

File: HashTable/349_IntersectionOfTwoArrays.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# @Author: xuezaigds@gmail.com
# @Last Modified time: 2016-07-01 21:00:38


class Solution(object):
    def intersection(self, nums1, nums2):
        return list(set(nums1) & set(nums2))


class Solution_2(object):
    def intersection(self, nums1, nums2):
        ans = []
        nums1_dict = {}
        for n in nums1:
            nums1_dict[n] = nums1_dict.get(n, 0) + 1

        for n in nums2:
            if n in nums1_dict and nums1_dict[n] != -1:
                ans.append(n)
                nums1_dict[n] = -1
        return ans


class Solution_3(object):
    def intersection(self, nums1, nums2):
        # sort the two list, and use two pointer to search to find common elements.
        ans = []
        nums1.sort()
        nums2.sort()
        i = j = 0
        while (i < len(nums1) and j < len(nums2)):
            if nums1[i] > nums2[j]:
                j += 1
            elif nums1[i] < nums2[j]:
                i += 1
            else:
                if not len(ans) or nums1[i] != ans[-1]:
                    ans.append(nums1[i])
                i += 1
                j += 1

        return ans

"""
[]
[]
[-1]
[]
[1,2,2,2,10,5]
[1,3,4,5,9]
"""
